### ISSUE TITLE
iso_oscar_gap: Check arguments (optionally)

### ISSUE THREAD

**Issue by thofma** [oscar-system/Oscar.jl#1066]
At the moment one can throw everything into the function returned by `iso_oscar_gap` which has the correct type. But it is not checked whether the argument is in the domain.
```julia
julia> F = GF(2)
Galois field with characteristic 2

julia> Fiso = Oscar.iso_oscar_gap(F)
Map from
Galois field with characteristic 2 to GAP: GF(2) defined by a julia-function with inverse

julia> Fiso(F(1))
GAP: Z(2)^0

julia> F5 = GF(5)
Galois field with characteristic 5

julia> Fiso(F(1)) # this should not work
GAP: Z(2)^0
```
The example is wrong... `Fiso(F(1))` is correct as is, `Fiso(F5(1))` should fail...

This can be done on the julia side, so there are two questions:
- Should the check be there by default? `@assert` would be a pointer comparison + throwing. Is this a performance penalty we want to take? I can do some timings if there is interest.
- If the check is not there by default, should the function have the `check::Bool` keyword argument?

CC: @fieker @ThomasBreuer @fingolfin

**Comment by fingolfin** [comment]
Indeed, this would be good to find bugs. A way to turn it off as well, so that when e.g. converting a matrix we don't have to repeat the check for each argument.

So perhaps have a `check::Bool` argument, though I'd tend to default `check` to `true`...

**Comment by ThomasBreuer** [comment]
I understand the idea as follows:
`Oscar.iso_oscar_gap` gets a `check` argument.
The methods of `Oscar.iso_oscar_gap` eventually create two functions that get wrapped in a mapping,
one from Oscar objects to GAP objects and one from GAP objects to Oscar objects.
The check is performed in both cases on the Oscar side, because it means just a parent check which is cheap.
Is there a cheaper way to implement this feature than calling `if check` inside the functions?
(Of course one could create different functions, depending on `check`, then no `if` is executed at runtime.)

**Comment by thofma** [comment]
Ah, so you were thinking about
```julia
Fiso = Oscar.iso_oscar_gap(F, check = false)
```
(Let's call this proposal (B)).

I actually had the following in mind:
```julia
Fiso = Oscar.iso_oscar_gap(F, check = bla)
y = Fiso(x, check = false)
```
(Let's call that proposal (A)).

In don't mind either way. For (A) I don't know a cheaper way than doing the `if check` at runtime. For (B) I would just branch before creating the function as you suggest.

**Comment by ThomasBreuer** [comment]
Thanks @thofma 

(A) means that one can decide about checking for each individual call of the functions that describe the mapping.
(Then the `check` argument for `Oscar.iso_oscar_gap` is intended as a default?)

(B) means that it is a property of the mapping whether one wants to check the calls of its functions (either check all or none).

Since you thought about (A), should I implement (A)?

**Comment by thofma** [comment]
Let's hear what @fingolfin and @fieker think

**Comment by fingolfin** [comment]
I can live with either approach. What I care about more is consistency with other types of maps. I.e. whatever we pick, ideally we do the same elsewhere, too (where it makes sense).

**Comment by fieker** [comment]
I fail to see the point of not poing the parent check - in fact a lot of the "generic" map code will check the parent anyway, I am happy to allw the paramter check, but a parent check is cheap, so optimizing one if statement  away by introducing another if statment sounds not right.
In situations where input validation is expensive: yes, but here the parent check shoudl always be done

**Comment by ThomasBreuer** [comment]
O.k., I will create a pull request in this direction:
- Let the two functions of the `MapFromFunc` support the optional `check` argument. (Just now I noticed that then the change from thofma/Hecke.jl/pull/592 does not really help, since we cannot supply `fmpz` etc. as a function because we want to add the parent check.)
- The default for `check` is `true`, and we do not guarantee that entering `false` will really cause that all checks are omitted.
- There can be situations where a check on the GAP side makes sense; these checks are not as cheap as a parent check on the Oscar side, thus omitting GAP side checks makes sense if `check = false` is given.

**Comment by lgoettgens** [comment]
I think this can be closed as of #1092, right?
